x/virtualgroup/keeper/keeper.go

```diff
@@ -647,3 +647,283 @@ func (k Keeper) CompleteSwapOut(ctx sdk.Context, gvgFamilyID uint32, gvgIDs []ui
 	}
 	return nil
 }
+
+func (k Keeper) SwapIn(ctx sdk.Context, gvgFamilyID uint32, gvgID uint32, successorSPID uint32, targetSP *sptypes.StorageProvider) error {
+	curTime := uint64(ctx.BlockTime().Unix())
+	// swapIn a family, the target sp needs to be exiting status if swapIn the family's as primary SP.
+	if gvgFamilyID != types.NoSpecifiedFamilyId {
+		if targetSP.Status != sptypes.STATUS_GRACEFUL_EXITING && targetSP.Status != sptypes.STATUS_FORCE_EXITING {
+			return sptypes.ErrStorageProviderWrongStatus.Wrapf("The target sp is not exiting, can not be swapped")
+		}
+		family, found := k.GetGVGFamily(ctx, gvgFamilyID)
+		if !found {
+			return types.ErrGVGFamilyNotExist
+		}
+		if family.PrimarySpId != targetSP.Id {
+			return types.ErrSwapInFailed.Wrapf("the family(ID: %d) primary SP(ID: %d) does not match the target SP(ID: %d) which need to be swapped", family.Id, family.PrimarySpId, targetSP.Id)
+		}
+		return k.setSwapInInfo(ctx, types.GetSwapInFamilyKey(gvgFamilyID), successorSPID, targetSP.Id, curTime)
+	}
+
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		return types.ErrGVGNotExist
+	}
+	if gvg.PrimarySpId == successorSPID {
+		return types.ErrSwapInFailed.Wrapf("The SP(ID=%d) is already the primary SP of GVG(ID=%d)", successorSPID, gvgID)
+	}
+	exist := false
+	for _, sspID := range gvg.GetSecondarySpIds() {
+		if sspID == successorSPID {
+			return types.ErrSwapInFailed.Wrapf("The sp(ID: %d) is already one of the secondary in this GVG(ID:%d)", successorSPID, gvgID)
+		}
+		if sspID == targetSP.Id {
+			exist = true
+		}
+	}
+	if !exist {
+		return types.ErrSwapInFailed.Wrapf("The sp(ID: %d) that needs swap out is not one of the secondary sps of gvg gvg(%s).", targetSP.Id, gvg.String())
+	}
+
+	// swap into GVG when there is a secondary SP is exiting.
+	if targetSP.Status == sptypes.STATUS_GRACEFUL_EXITING || targetSP.Status == sptypes.STATUS_FORCE_EXITING {
+		return k.setSwapInInfo(ctx, types.GetSwapInGVGKey(gvgID), successorSPID, targetSP.Id, curTime)
```

Comment:
return here will skip redundancy check. We can move this common check before line 654, to reduce duplicated codes.

Reply:
If a GVG has broke the redundancy, say {1,1,2,3,4,5,6}, 1 is primary SP will be restricted to exit, any other secondary SP is allowed to exit.